A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,398 @@
+[load_balancing_policy][] field used for custom LB policies which means custom
+LB policies and stateful session affinity features will be incompatible with
+each other. This will need to be fixed eventually to be able to use these
+two features together.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing and just pass the RPC through (to the next
+filter). For example if the `path` configuration value is `/Package1.Service2`
+and the RPC method is `/Package2.Service1/Method3` then just pass the RPC
+through. Note, for path matching the rules described in
+[rfc6265 section-5.1.4][rfc6265-path-match] will apply.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
+warning and pass the RPC through. For a valid `override-host` value set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path of the RPC, if the RPC path does not match the
+  `path` config value of the filter then skip further processing and
+  pass the response through. Otherwise check if `upstreamAddress` is
```

Comment:
`upstreamAddress` is set from the value of the cookie. If there was no cookie then `upstreamAddress` is not set.

Reply:
Okay this is a bit complicated although I agree the text needs to be improved.

There can be a route based filter_config_override because of which this filter may be skipped if the RPC path does not match the path set in the `routeConfiguration` - that is automatic (or already implemented in gRPC) so does not need to be stated.

However there is also a path inside the Cookie config which can only be checked by the filter once the filter is invoked (which means the RPC path matched the applicable `routeConfiguration`). The filter can then decide to skip further processing because of that mismatch. And in that case too, the corresponding response processing needs to be skipped. But I realized that at least in Java the interceptor (aka filter) can decide to not invoke the `ResponseListener` in this case and that means no need to check any `upstreamAddress` etc.

I'll modify the text accordingly.